cpp/ql/src/semmle/code/cpp/ir/internal/CppType.qll

```diff
@@ -0,0 +1,500 @@
+private import cpp
+private import semmle.code.cpp.Print
+private import semmle.code.cpp.ir.implementation.IRType
+private import semmle.code.cpp.ir.implementation.raw.internal.IRConstruction as IRConstruction
+
+/**
+ * Works around an extractor bug where a function reference gets a size of one byte.
+ */
+private int getTypeSizeWorkaround(Type type) {
+  exists(Type unspecifiedType |
+    unspecifiedType = type.getUnspecifiedType() and
+    (
+      unspecifiedType instanceof FunctionReferenceType and
+      result = any(NullPointerType t).getSize()
+      or
+      exists(PointerToMemberType ptmType |
+        ptmType = unspecifiedType and
+        (
+          if ptmType.getBaseType().getUnspecifiedType() instanceof RoutineType
+          then result = any(NullPointerType t).getSize() * 2
+          else result = any(NullPointerType t).getSize()
+        )
+      )
+    )
+  )
+}
+
+private int getTypeSize(Type type) {
+  if exists(getTypeSizeWorkaround(type))
+  then result = getTypeSizeWorkaround(type)
+  else result = type.getSize()
+}
+
+/**
+ * Holds if an `IRBooleanType` with the specified `byteSize` should exist.
+ */
+predicate hasBooleanType(int byteSize) {
+  byteSize = getTypeSize(any(BoolType type))
+}
+
+private predicate isSigned(IntegralOrEnumType type) {
+  type.(IntegralType).isSigned() or
+  exists(Enum enumType |
+    enumType = type.getUnspecifiedType() and
+    (
+      enumType.getExplicitUnderlyingType().getUnspecifiedType().(IntegralType).isSigned() or
+      not exists(enumType.getExplicitUnderlyingType())  // Assume signed.
```

Comment:
Why assume signed? GCC enums are unsigned unless one of the constants is negative: https://gcc.gnu.org/onlinedocs/gcc/Structures-unions-enumerations-and-bit-fields-implementation.html. I'm fine with having this assumption be imperfect as long as we leave a note about it.

Reply:
Sounds like we need an extractor change to get this right. I'll add a comment to the workaround for now. I might be able to implement a better workaround by looking at the type of the enum constants themselves, depending on how the extractor models them.

Reply:
I've changed the default to unsigned, and added a comment.